src/renderer/services/file-management-system/index.ts

```diff
@@ -398,85 +397,134 @@ export default class FileManagementSystem {
-          }
-          const deserailizedMd5Hasher = await Md5Hasher.deserialize(partiallyCalculatedMd5);
-          await this.fss.finalize(fssUploadId, deserailizedMd5Hasher.digest());
-        } else if (fssStatus.status === UploadStatus.WORKING) {
-          const { originalPath } = upload.serviceFields.files[0].file;
-          const { size: fileSize } = await fs.promises.stat(originalPath);
-          await this.uploadInChunks({
-            fssUploadId,
-            source: originalPath,
-            chunkSize: fssStatus.chunkSize,
-            user: upload.user,
-            onProgress: (bytesUploaded) =>
-              onProgress(upload.jobId, { bytesUploaded, totalBytes: fileSize }),
-            initialChunkNumber: lastChunkNumber,
-            partiallyCalculatedMd5
-          });
-        }
+  private async retryChunkedUpload(
+    fssStatus: UploadStatusResponse
+  ) {
+    const [lastChunkNumber, partiallyCalculatedMd5] = await this.getChunkedUploadProgress(fssStatus);
+    if (!partiallyCalculatedMd5) {
+      throw new Error('No partial MD5 for chunk ' + lastChunkNumber);
+    }
+    const deserailizedMd5Hasher = await Md5Hasher.deserialize(partiallyCalculatedMd5);
+    await this.fss.retryFinalizeMd5(fssStatus.uploadId, deserailizedMd5Hasher.digest());
+  }
+
+  private async resumeUploadInChunks(
+    upload: UploadJob,
+    fssStatus: UploadStatusResponse,
+    onProgress: (uploadId: string, progress: UploadProgressInfo) => void
+  ) {
+    const { originalPath } = upload.serviceFields.files[0].file;
+    const { size: fileSize } = await fs.promises.stat(originalPath);
+    const [lastChunkNumber, partiallyCalculatedMd5] = await this.getChunkedUploadProgress(fssStatus);
+
+    await this.uploadInChunks({
+      fssStatus,
+      source: originalPath,
+      user: upload.user,
+      onProgress: (bytesUploaded) =>
+        onProgress(upload.jobId, { bytesUploaded, totalBytes: fileSize }),
+      initialChunkNumber: lastChunkNumber,
+      partiallyCalculatedMd5
+    });
+  }
+
+  private async updateProgress(
+    fssUploadId: string,
+    fileSize: number,
+    onProgress: (progress: UploadProgressInfo) => void) {
+    const fssStatusResponse = await this.fss.getStatus(fssUploadId);
+    let fssStatus = fssStatusResponse?.status;
+    while (fssStatus !== UploadStatus.COMPLETE) {
+      await FileManagementSystem.sleep(1000); //TODO too short?  Tests timeout if > 2 sec
+      const fssStatusResponse = await this.fss.getStatus(fssUploadId);
+      fssStatus = fssStatusResponse?.status;
+      switch (fssStatus) {
+        case UploadStatus.WORKING:
+          onProgress({ bytesUploaded: fssStatusResponse.currentFileSize, totalBytes: fileSize });
+          break;
+        case UploadStatus.POST_PROCESSING:
+          onProgress({ bytesUploaded: fssStatusResponse.currentFileSize, totalBytes: fileSize });
+          break;
+        case UploadStatus.INACTIVE:
+        case UploadStatus.RETRY:
```

Comment:
Based on the logic in the resume function I believe we can assume the finalize step needs to be retried (I have another comment double checking that). Since that is actionable by the user, might be worth changing the error statement a little for the retry case.